File: v6/prolog/labs/rel_spreading/probes/go/c4_width_subtyping.go
// C4 width subtyping in Go: is the wider (embedding) struct accepted where
// the narrower one is wanted? EXPECTED: refused for struct types.
package main

import "fmt"

type Narrow struct {
	Id int64
}

type Wide struct {
	Narrow
	Extra int64
}

func takesNarrow(row Narrow) {
	fmt.Println(row.Id)
}

func main() {
	w := Wide{Narrow{1}, 2}

	// (1) the embedding struct where the embedded one is wanted: refused
	takesNarrow(w)

	// (2) the promoted field must be named explicitly
	takesNarrow(w.Narrow)

	// (3) a structurally identical but separately declared struct: refused
	type AlsoNarrow struct {
		Id int64
	}
	takesNarrow(AlsoNarrow{1})
}

File: v6/prolog/labs/rel_spreading/probes/rust/c4_width_subtyping.rs
// C4 width subtyping in Rust: is a wider struct accepted where the narrower
// one is wanted? EXPECTED: refused, struct types are nominal.

struct Narrow {
    id: i64,
}

struct Wide {
    id: i64,
    extra: i64,
}

fn takes_narrow(_row: Narrow) {}

fn main() {
    let wide = Wide { id: 1, extra: 2 };
    takes_narrow(wide);

    // even a structurally IDENTICAL second struct is a different type
    struct AlsoNarrow {
        id: i64,
    }
    let also = AlsoNarrow { id: 1 };
    takes_narrow(also);

    // tuples are structural, and there width is exact
    fn takes_two(_t: (i64, i64)) {}
    takes_two((1, 2, 3));
}
